### PR TITLE
feat(amf): Support for Protocol Configuration Options IE

### PR DESCRIPTION
**Pull request by sreedharkumartn**
Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

## Summary
In this PR, Issues related to VoNR are fixed.
- Added IE in registration accept to support for 5GS network feature support. 
- Support for ```P_CSCF_IPV4_ADDRESS_REQUEST``` is added in Protocol Configuration Options IE.
- Fixed SST miss match in service request message.
- UE Aggregate bit rate IE is included in initial context setup request at service request only.
- Fixed the Qos Flow identifier issue.
- Fixed the DQR field in QOS Rules IE, E-Bit field in QOS Flow Identifier IE at PDU Session accept message.

## Test Plan
- Verified with Unit Test.
image

- Verified with spirent setup.
1. For ```P_CSCF_IPV4_ADDRESS_REQUEST``` in PDU Session request message.
image
2. In PDU Session accept message will get ```P_CSCF_IPV4_ADDRESS```.
image

- Done basic sanity (basic registration, pdu session accept, traffic flow) with UERANSIM. Please find the PCAP, mme log and syslog in below zip file.
[Basic_sanity.zip](https://github.com/magma/magma/files/9444405/Basic_sanity.zip)

- Done Service request testing with UERANSIM. Please find the PCAP, mme log and syslog in below zip file.
[Service_request.zip](https://github.com/magma/magma/files/9444431/Service_request.zip)

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Review by ganeshg87 — APPROVED** [review]
lgtm

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GNetworkFeatureSupport.cpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+#include <stdlib.h>
+#include <stdint.h>
+#ifdef __cplusplus
+extern "C" {
+#endif
+#include "lte/gateway/c/core/oai/common/log.h"
+#ifdef __cplusplus
+}
+#endif
+#include "lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GNetworkFeatureSupport.hpp"
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GNetworkFeatureSupport.cpp]: Put the header at start in accordance to link, https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GNetworkFeatureSupport.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GNetworkFeatureSupport.cpp:14]: Leave the space

**Review by LKreutzer — APPROVED** [review]
Bazel changes are good

I tested the test changes for flakiness and all 1000 runs were successfull in all configs: 
- `bazel test //lte/gateway/c/core/oai/test/amf:amf_encode_decode_test --runs_per_test=1000`
- `bazel test //lte/gateway/c/core/oai/test/amf:amf_encode_decode_test --runs_per_test=1000 --config=asan`
- `bazel test //lte/gateway/c/core/oai/test/amf:amf_encode_decode_test --runs_per_test=1000 --config=production`

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GNetworkFeatureSupport.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  return decoded;
+}
+
+int NetworkFeatureSupportMsg::EncodeNetworkFeatureSupportMsg(
+    NetworkFeatureSupportMsg* networkfeature, uint8_t iei, uint8_t* buffer,
+    uint32_t len) {
+  uint32_t encoded = 0;
+
+  /*
+   * Checking IEI and pointer
```
Comment (rsarwad) [inline lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GNetworkFeatureSupport.cpp]: Put in single line comment

Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GNetworkFeatureSupport.cpp]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GNetworkFeatureSupport.cpp (hunk 2 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GNetworkFeatureSupport.cpp:14]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GNetworkFeatureSupport.cpp (hunk 1 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/nas5g/src/ies/M5GNetworkFeatureSupport.cpp]: Addressed Review Comment

**Review by rsarwad — APPROVED** [review]
lgtm

File: lte/gateway/c/core/oai/tasks/amf/amf_as.cpp [hunk 4]
```diff
@@ -498,6 +498,26 @@ int amf_reg_acceptmsg(const guti_m5_t* guti, const tai_t* tai,
...
+      .network_feature.MPSI = 0;
+  nas_msg->security_protected.plain.amf.msg.registrationacceptmsg
+      .network_feature.IWK_N26 = 0;
+  nas_msg->security_protected.plain.amf.msg.registrationacceptmsg
+      .network_feature.EMF = 0;
+  nas_msg->security_protected.plain.amf.msg.registrationacceptmsg
+      .network_feature.EMC = 0;
+  nas_msg->security_protected.plain.amf.msg.registrationacceptmsg
+      .network_feature.IMS_VoPS_N3GPP = 0;
+  nas_msg->security_protected.plain.amf.msg.registrationacceptmsg
```
Comment (panyogesh) [inline lte/gateway/c/core/oai/tasks/amf/amf_as.cpp:515]: Make the default as 0.

File: lte/gateway/c/core/oai/tasks/amf/amf_as.cpp [hunk 5]
```diff
@@ -1278,6 +1298,12 @@ status_code_e initial_context_setup_request(amf_ue_ngap_id_t ue_id,
     }
   }
```
Comment (panyogesh) [inline lte/gateway/c/core/oai/tasks/amf/amf_as.cpp:1300]: put some additional comments,

File: lte/gateway/c/core/oai/tasks/amf/amf_as.cpp (hunk 4 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/amf/amf_as.cpp:515]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/amf/amf_as.cpp (hunk 5 above)
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/amf/amf_as.cpp:1300]: Addressed Review Comment

File: lte/gateway/c/core/oai/tasks/amf/amf_as.cpp [hunk 6]
```diff
@@ -1278,6 +1298,14 @@ status_code_e initial_context_setup_request(amf_ue_ngap_id_t ue_id,
...
 
+  // UE Aggregate bit rate IE will be added only if PDU Session Resource setup
+  // transfer IE is added in intial context setup request message.
```
[the remark this answers is not in the document]
Comment (sreedharkumartn) [reply lte/gateway/c/core/oai/tasks/amf/amf_as.cpp]: Addressed

**Pull request merged**